apiserver/facades/agent/upgrader/upgrader.go

```diff
@@ -91,43 +95,59 @@ func NewUpgraderAPI(
 }
 
 // WatchAPIVersion starts a watcher to track if there is a new version
-// of the API that we want to upgrade a machine to.
+// of the API that we want to upgrade an application, machine, model or
+// unit to.
 func (u *UpgraderAPI) WatchAPIVersion(ctx context.Context, args params.Entities) (params.NotifyWatchResults, error) {
 	result := params.NotifyWatchResults{
 		Results: make([]params.NotifyWatchResult, len(args.Entities)),
 	}
 	for i, agent := range args.Entities {
-		tag, err := names.ParseMachineTag(agent.Tag)
+		tag, err := names.ParseTag(agent.Tag)
 		if err != nil {
-			return params.NotifyWatchResults{}, errors.Trace(err)
+			result.Results[i].Error = apiservererrors.ParamsErrorf(
+				params.CodeTagInvalid, "%s", err,
+			)
+			continue
 		}
+
 		tagID := tag.Id()
-		upgraderAPIWatcher, err := u.modelAgentService.WatchMachineTargetAgentVersion(ctx, machine.Name(tagID))
+		var upgraderAPIWatcher watcher.NotifyWatcher
+
+		switch tag.Kind() {
+		case names.ControllerTagKind, names.ModelTagKind:
+			upgraderAPIWatcher, err = u.modelAgentService.WatchModelTargetAgentVersion(ctx)
+		case names.MachineTagKind:
+			upgraderAPIWatcher, err = u.modelAgentService.WatchMachineTargetAgentVersion(ctx, machine.Name(tagID))
+		case names.UnitTagKind:
+			// Used in kubernetes models.
+			upgraderAPIWatcher, err = u.modelAgentService.WatchUnitTargetAgentVersion(ctx, tagID)
+		default:
+			result.Results[i].Error = apiservererrors.ParamsErrorf(
+				params.CodeNotValid, "%s", tag.String(),
+			)
+			continue
+		}
 
 		switch {
-		case errors.Is(err, errors.NotValid):
+		case errors.Is(err, modelerrors.NotFound), errors.Is(err, applicationerrors.ApplicationNotFound),
```

Comment:
Why are we expecting an application error for a not found unit?
In fact, the existing code is wrong, both the doc string and the error returned
```
// WatchUnitTargetAgentVersion is responsible for watching the target agent
// version for unit and reporting when there has been a change via a
// [watcher.NotifyWatcher]. The following errors can be expected:
// - [applicationerrors.NotFound] - When no unit exists for the provided name.
// - [modelerrors.NotFound] - When the model of the unit no longer exists.
func (s *ModelService) WatchUnitTargetAgentVersion(
	ctx context.Context,
	unitName string,
) (watcher.NotifyWatcher, error) {
	err := s.st.CheckUnitExists(ctx, unitName)
	if errors.Is(err, applicationerrors.UnitNotFound) {
		return nil, errors.Errorf(
			"unit %q does not exist", unitName,
		).Add(machineerrors.MachineNotFound)
```

Reply:
This is my mistake. Bad copy and paste bug.